tests/unit/valgrind.suppressions

```diff
@@ -10,6 +12,50 @@
    fun:pthread_create@@GLIBC_2.2.5
    fun:main
 }
+{
+   pthread_false_positive_glibc_2_34
+   Memcheck:Leak
+   match-leak-kinds: possible
+   fun:calloc
+   fun:calloc
+   fun:allocate_dtv
+   fun:_dl_allocate_tls
+   fun:allocate_stack
+   fun:pthread_create@@GLIBC_2.34
+   ...
+   fun:main
+}
+
+# > backtrace() and backtrace_symbols_fd() don't call malloc()
+# > explicitly, but they are part of libgcc, which gets loaded
+# > dynamically when first used.  Dynamic loading usually triggers
+# > a call to [malloc(3)](https://man7.org/linux/man-pages/man3/malloc.3.html).
+# https://man7.org/linux/man-pages/man3/backtrace_symbols_fd.3.html
```

Comment:
The full trace of the leak looks like:
```
agent-1  | ==6495== 2,304 bytes in 1 blocks are possibly lost in loss record 8 of 9
agent-1  | ==6495==    at 0x4848899: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
agent-1  | ==6495==    by 0x4005D97: malloc (rtld-malloc.h:56)
agent-1  | ==6495==    by 0x4005D97: _dlfo_mappings_segment_allocate (dl-find_object.c:217)
agent-1  | ==6495==    by 0x4005D97: _dl_find_object_update_1 (dl-find_object.c:671)
agent-1  | ==6495==    by 0x4005D97: _dl_find_object_update (dl-find_object.c:804)
agent-1  | ==6495==    by 0x400ECCF: dl_open_worker_begin (dl-open.c:735)
agent-1  | ==6495==    by 0x49D8A97: _dl_catch_exception (dl-error-skeleton.c:208)
agent-1  | ==6495==    by 0x400DF99: dl_open_worker (dl-open.c:782)
agent-1  | ==6495==    by 0x49D8A97: _dl_catch_exception (dl-error-skeleton.c:208)
agent-1  | ==6495==    by 0x400E34D: _dl_open (dl-open.c:883)
agent-1  | ==6495==    by 0x49D8E00: do_dlopen (dl-libc.c:95)
agent-1  | ==6495==    by 0x49D8A97: _dl_catch_exception (dl-error-skeleton.c:208)
agent-1  | ==6495==    by 0x49D8B62: _dl_catch_error (dl-error-skeleton.c:227)
agent-1  | ==6495==    by 0x49D8F36: dlerror_run (dl-libc.c:45)
agent-1  | ==6495==    by 0x49D8F36: __libc_dlopen_mode (dl-libc.c:162)
agent-1  | ==6495==    by 0x4989610: __libc_unwind_link_get (unwind-link.c:50)
agent-1  | ==6495==    by 0x4989610: __libc_unwind_link_get (unwind-link.c:40)
agent-1  | ==6495==    by 0x4997B76: backtrace (backtrace.c:69)
agent-1  | ==6495==    by 0x15C1B3: s2n_calculate_stacktrace (s2n_errno.c:457)
agent-1  | ==6495==    by 0x15ABD4: raises_error (s2n_stacktrace_test.c:22)
agent-1  | ==6495==    by 0x159806: main (s2n_stacktrace_test.c:38)
agent-1  | ==6495== 
```

I think this leak is platform depends. The old Valgrind tests run on Ubuntu 18,and they don't produce this memory issue. However, for Ubuntu 22 and 24, this issue will be triggered. I tried `pthread_exit(NULL)`, but that doesn't really help with this problem.

The man page for `backtrace()` says we need to load `libgcc` before we call `backtrace()`. https://man7.org/linux/man-pages/man3/dlopen.3.html

I am trying to do `void *handle = dlopen("libgcc_s.so.1", RTLD_NOW)` and `dlclose(handle)`. That hasn't solved the issue. There is another related issue I found: https://stackoverflow.com/questions/78354256/how-to-detect-if-libgcc-has-been-loaded

Reply:
As discussed offline, none of the above proves that the memory leak isn't a real problem with our code. Can you update this thread with your latest evidence that the valgrind error can be safety suppressed?

Reply:
The reason for this memory issue to be triggered is because `backtrace()` needs to dynamically load `libgcc` when it was first called. My theory is this possible loss of memory is not scaleable, since `libgcc` will only needs to be loaded once in the program's lifetime. To test this theory, I write a for loop which runs 100 loops in `s2n_stacktrace_test` and see if the possible memory loss increases 100 times.
```
#include "api/s2n.h"
#include "s2n_test.h"
#include "utils/s2n_blob.h"

#define TEST_LOOP 100

int raises_error()
{
    POSIX_BAIL(S2N_ERR_INVALID_ARGUMENT);
}

int main(int argc, char **argv)
{
    BEGIN_TEST();
    EXPECT_SUCCESS(s2n_disable_tls13_in_test());
#ifdef S2N_STACKTRACE
    EXPECT_SUCCESS(s2n_stack_traces_enabled_set(true));
    struct s2n_stacktrace trace;
    /* If nothing has errored yet, we have no stacktrace */
    EXPECT_SUCCESS(s2n_get_stacktrace(&trace));
    EXPECT_NULL(trace.trace);
    EXPECT_EQUAL(trace.trace_size, 0);

    for (uint8_t i = 0; i < TEST_LOOP; i ++) {
        /* Raise an error, and see that it generates a stacktrace */
        EXPECT_FAILURE(raises_error());
        EXPECT_SUCCESS(s2n_get_stacktrace(&trace));
        EXPECT_NOT_NULL(trace.trace);
        EXPECT_NOT_EQUAL(trace.trace_size, 0);

        /* Test printing the stacktrace. */
        FILE *stream = fopen("/dev/null", "w");
        EXPECT_SUCCESS(s2n_print_stacktrace(stream));
        fclose(stream);

        /* Free the stacktrace to avoid memory leaks */
        EXPECT_SUCCESS(s2n_free_stacktrace());
    }
    END_TEST();
#else
    END_TEST();
#endif
}
```

I ran valgrind against this updated test, and the result looks like this
```
ubuntu@ip-172-31-59-74:~/s2n-tls/tests/unit$ valgrind             --leak-check=full             --leak-resolution=high             --trace-children=yes             --run-libc-freeres=no             --error-exitcode=123             --error-limit=no             --num-callers=40             --undef-value-errors=no             --suppressions=valgrind.suppressions ../../build/bin/s2n_stacktrace_test 
==522635== Memcheck, a memory error detector
==522635== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==522635== Using Valgrind-3.22.0 and LibVEX; rerun with -h for copyright info
==522635== Command: ../../build/bin/s2n_stacktrace_test
==522635== 
Running /home/ubuntu/s2n-tls/tests/unit/s2n_stacktrace_test.c ... PASSED        805 tests
==522635== 
==522635== HEAP SUMMARY:
==522635==     in use at exit: 5,726 bytes in 9 blocks
==522635==   total heap usage: 29,235 allocs, 29,226 frees, 2,847,370 bytes allocated
==522635== 
==522635== 2,304 bytes in 1 blocks are possibly lost in loss record 9 of 9
==522635==    at 0x4846828: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==522635==    by 0x4004EDC: malloc (rtld-malloc.h:56)
==522635==    by 0x4004EDC: _dlfo_mappings_segment_allocate (dl-find_object.c:217)
==522635==    by 0x4004EDC: _dl_find_object_update_1 (dl-find_object.c:671)
==522635==    by 0x4004EDC: _dl_find_object_update (dl-find_object.c:805)
==522635==    by 0x400DC37: dl_open_worker_begin (dl-open.c:756)
==522635==    by 0x400151B: _dl_catch_exception (dl-catch.c:237)
==522635==    by 0x400CD1F: dl_open_worker (dl-open.c:803)
==522635==    by 0x400151B: _dl_catch_exception (dl-catch.c:237)
==522635==    by 0x400D163: _dl_open (dl-open.c:905)
==522635==    by 0x49E80D4: do_dlopen (dl-libc.c:95)
==522635==    by 0x400151B: _dl_catch_exception (dl-catch.c:237)
==522635==    by 0x4001668: _dl_catch_error (dl-catch.c:256)
==522635==    by 0x49E84EE: dlerror_run (dl-libc.c:45)
==522635==    by 0x49E84EE: __libc_dlopen_mode (dl-libc.c:162)
==522635==    by 0x498D3C0: __libc_unwind_link_get (unwind-link.c:51)
==522635==    by 0x4999E46: backtrace (backtrace.c:69)
==522635==    by 0x15D2DB: s2n_calculate_stacktrace (s2n_errno.c:457)
==522635==    by 0x15BCFE: raises_error (s2n_stacktrace_test.c:24)
==522635==    by 0x15A90D: main (s2n_stacktrace_test.c:41)
==522635== 
==522635== LEAK SUMMARY:
==522635==    definitely lost: 0 bytes in 0 blocks
==522635==    indirectly lost: 0 bytes in 0 blocks
==522635==      possibly lost: 2,304 bytes in 1 blocks
==522635==    still reachable: 3,422 bytes in 8 blocks
==522635==         suppressed: 0 bytes in 0 blocks
==522635== Reachable blocks (those to which a pointer was found) are not shown.
==522635== To see them, rerun with: --leak-check=full --show-leak-kinds=all
==522635== 
==522635== For lists of detected and suppressed errors, rerun with: -s
==522635== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

Compare this result as the one that I posted in the previous comment, we can see that this test passes 805 tests, because of the loop, and the amount of possible loss stays at 2204 bytes. Furthermore, the location where triggers the error stays the same as well. Hence, I think this issue is not scaleable and can be suppressed at this moment.

Reply:
This is an example of memory leak that does scale:
Remove `DEFER_CLEANUP` macro in `s2n_x509_validator_test.c`. Run it by one time, and then run that test for 100 times:
https://github.com/boquan-fang/s2n-tls/blob/606390394913cdc724b4b0e18daad78bae092f47/tests/unit/s2n_x509_validator_test.c#L202-L227

The result for the first run is:
```
==6774== 2,447 bytes in 1 blocks are definitely lost in loss record 7 of 8
```
The result for the second run is:
```
==6655== 244,700 bytes in 100 blocks are definitely lost in loss record 42 of 42
```
This example shows that this memory leak scale 100 times. The codebuild job for such loss can be found in [job1](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/Valgrind/build/Valgrind%3A41241a06-f047-4f80-a2c3-5e846c8e1a72?region=us-west-2) and [job2](https://us-west-2.console.aws.amazon.com/codesuite/codebuild/024603541914/projects/Valgrind/build/Valgrind%3Aa8753558-ae28-4480-b3d9-0997d96485ab?region=us-west-2).